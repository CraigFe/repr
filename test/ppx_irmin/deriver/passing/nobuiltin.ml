(* When a type is annotated with the [nobuiltin] annotation, it should be
   considered as an abstract type (i.e. don't pull representations from
   [Irmin_type.Type]). *)

type unit = string [@@deriving irmin]

(* Shadow [Stdlib.unit] *)
module Nobuiltin_t = struct
  type t = (unit[@nobuiltin]) [@@deriving irmin]

  (* [t]'s repr should be for strings. *)
  let (_ : string Irmin_type.Type.t) = t
end

module Nobuiltin_foo = struct
  type foo = (unit[@irmin.nobuiltin]) [@@deriving irmin]

  (* [foo]'s repr should be for strings too. *)
  let (_ : string Irmin_type.Type.t) = foo_t
end

module Nobuiltin_operator = struct
  (* Define our own representation of [result]. *)
  let result_t a b = Irmin_type.Type.pair a b

  let int32_t = Irmin_type.Type.int

  let int64_t = Irmin_type.Type.bool

  type u = (((int32[@nobuiltin]), int64) result[@nobuiltin]) [@@deriving irmin]

  let (_ : (int * int64) Irmin_type.Type.t) = u_t
end
