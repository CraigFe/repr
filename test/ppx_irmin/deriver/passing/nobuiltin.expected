type unit = string[@@deriving irmin]
include struct let unit_t = Irmin_type.Type.string end[@@ocaml.doc "@inline"]
[@@merlin.hide ]
module Nobuiltin_t =
  struct
    type t = ((unit)[@nobuiltin ])[@@deriving irmin]
    include struct let t = unit_t end[@@ocaml.doc "@inline"][@@merlin.hide ]
    let (_ : string Irmin_type.Type.t) = t
  end
module Nobuiltin_foo =
  struct
    type foo = ((unit)[@irmin.nobuiltin ])[@@deriving irmin]
    include struct let foo_t = unit_t end[@@ocaml.doc "@inline"][@@merlin.hide
                                                                  ]
    let (_ : string Irmin_type.Type.t) = foo_t
  end
module Nobuiltin_operator =
  struct
    let result_t a b = Irmin_type.Type.pair a b
    let int32_t = Irmin_type.Type.int
    let int64_t = Irmin_type.Type.bool
    type u = (((((int32)[@nobuiltin ]), int64) result)[@nobuiltin ])[@@deriving
                                                                    irmin]
    include struct let u_t = result_t int32_t Irmin_type.Type.int64 end
    [@@ocaml.doc "@inline"][@@merlin.hide ]
    let (_ : (int * int64) Irmin_type.Type.t) = u_t
  end
